---
title: "Assignment 9 - BME4550: Data Science"
author: "James Gonzalez"
output: html_document
---

```{r setup, include=FALSE}

```

```{r}
knitr::opts_chunk$set(echo = TRUE)
library(devtools)
library(dagdata)
data(admissions)
library(GSE5859)
data(GSE5859)
```


###Exercises on page 419

1. Let’s compute the proportion of men who were accepted:
```{r}
    index = which(admissions$Gender==1)
    accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
    applied = sum(admissions$Number[index])
    accepted/applied
```
What is the proportion of women that were accepted?
```{r}
index = which(admissions$Gender==0)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
```

2. Now that we have observed different acceptance rates between genders, test for the significance of this result. If you perform an independence test, what is the p-value? 

```{r}
atab<-as.data.frame(matrix(0,2,2))
rownames(atab)<-c("Men","Women")
colnames(atab)<-c("Accepted","Rejected")
index = admissions$Gender==1
men=admissions[index,]
women=admissions[!index,]
atab[1,1]= sum(men$Number * men$Percent/100)
atab[1,2]= sum(men$Number*(1-men$Percent/100))
windex = which(admissions$Gender==0)
atab[2,1]= sum(women$Number * women$Percent/100)
atab[2,2]= sum(women$Number * (1-women$Percent/100))
atab
chisq.test(atab)$p.value
```

How can this be? This is referred to as Simpson’s Paradox. In the following questions we will try to decipher why this is happening.

```{r}
index = admissions$Gender==1
men = admissions[index,]
women = admissions[!index,]
print( data.frame( major=admissions[1:6,1],men=men[,3], women=women[,3]) )
```

3. We can quantify how “hard” a major is by using the percent of students that were accepted.
Compute the percent that were accepted (regardless of gender) to each major and call this vector H. Which is the hardest major?

```{r}
major = admissions[1:6,1]
men = admissions[1:6,]
women =admissions[7:12,]
H = (men$Number*men$Percent/100 + women$Number*women$Percent/100) / (men$Number+women$Number)
H
major[which.min(H)]
min(H)
```

Major F is the hardest.

4. What proportion is accepted for this major?

```{r}
min(H)
```

5. For men, what is the correlation between the number of applications across majors and H?

```{r}
cor(H,men$Number)
cor(men$Number,H)
```

The proportion is the same at about 0.765.

6. For women, what is the correlation between the number of applications across majors and H?

```{r}
cor(H,women$Number)
```

7. Given the answers to the above, which best explains the differences in admission percentages
when we combine majors?

• A) We made a coding mistake when computing the overall admissions percentages.
• B)There were more total numbers of women applications which made the denominator much bigger.
• C) There is confounding between gender and preference for “hard” majors: females are more likely to apply to harder majors.
• D)The sample size for the individual majors was not large enough to draw the correct conclusion.

There is a counfounding between gender and preference for "harder" majors: females are more likely to apply to harder majors.

###Exercises on page 431

```{r}
library(Biobase)
library(GSE5859)
data(GSE5859)
```

1. Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses, it does appear to have an effect so we will explore this here.
You can extract the year from each date like this:
 
```{r}
sampleInfo <- pData(e)
year = format(sampleInfo$date,"%y")
```

Note that ethnic group and year is almost perfectly confounded:

```{r}
table(year,sampleInfo$ethnicity)
```

2. For how many of these years do we have more than one ethnicity represented?

```{r}
tab=table(year,sampleInfo$ethnicity)
print(tab)
x=rowSums(tab!=0)
sum(x>=2)
```

3. Repeat the above exercise, but now, instead of year, consider the month as well.Specifically, instead of the year variable defined above use: 

```{r}
month.year = format(sampleInfo$date,"%m%y")
```

For what proportion of these month.year values do we have more than one ethnicity
represented?

```{r}
tab=table(year,sampleInfo$ethnicity)
print(tab)
x=rowSums(tab!=0)
sum(x>=2)
```

. Repeat the above exercise, but now, instead of year, consider the month as well.Specifically, instead of the year variable defined above use: 

```{r}
month.year = format(sampleInfo$date,"%m%y")
```

For what proportion of these month.year values do we have more than one ethnicity
represented?

```{r}
tab=table(year,sampleInfo$ethnicity)
print(tab)
x=rowSums(tab!=0)
sum(x>=2)
```

4. Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene.

How many genes have q-values < 0.05 ?

```{r}
library(qvalue)
library(genefilter)
geneExpression = exprs(e)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("02","03") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

5. What is the estimate of pi0 provided by qvalue:

```{r}
qval$pi0
```

6. Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those processed in 2004. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values less than 0.05?

```{r}
library(qvalue)
library(genefilter)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("03","04") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

7. Now we are going to compare ethnicities as was done in the original publication in which
these data were first presented. Use the qvalue function to compare the ASN population to the CEU population. Once again, use the qvalue function to obtain q-values.
How many genes have q-values < 0.05 ?

```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
index = which(ethnicity%in% c("CEU","ASN"))
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

8. Over 80% of genes are called differentially expressed between ethnic groups. However, due
to the confounding with processing date, we need to confirm these differences are actually due to ethnicity. This will not be easy due to the almost perfect confounding. However, above we noted that two groups were represented in 2005. Just like we stratified by majors to remove the “major effect” in our admissions example, here we can stratify by year and perform a t-test comparing ASN and CEU, but only for samples processed in 2005.

How many genes have q-values < 0.05 ?

```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
index = which(ethnicity%in% c("CEU","ASN") & year=="05")
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

Notice the dramatic drop in the number of genes with q-value < 0.05 when we fix the year. However, the sample size is much smaller in this latest analysis which means we have less power:

```{r}
table(sampleInfo$ethnicity[index])
```

9. To provide a more balanced comparison, we repeat the analysis, but now taking 3 random CEU samples from 2002. Repeat the analysis above, but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, set.seed(3)
How many genes have q-values < 0.05 ?

```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
indexASN = which(ethnicity%in% c("ASN") & year=="05")
indexCEU = which(ethnicity%in% c("CEU") & year=="02")
set.seed(3)
indexCEU<-sample(indexCEU,3)
index<-c(indexASN,indexCEU)
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

##Exercises on page 442

###1. Using the functions rowt tests and qvalue compare the two groups. Because this is a smaller dataset which decreases our power, we will use the more lenient FDR cut-off of 10%.

How many gene have q-values less than 0.1?

```{r, eval = FALSE}
sex = sampleInfo$group
month = factor( format(sampleInfo$date,"%m"))
library(qvalue)
library(genefilter)
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.1)
```

59 genes have q-values less than 0.1. 

###2. For the list calculated above, what proportion of this list is on chrX or chrY?
```{r, eval = FALSE}
library(qvalue)
library(genefilter)
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

The proportion of the list on chrX or chrY is 0.3389831.

###3. We can also check how many of the chromosomes X and Y genes we detected as different. How many are on Y?
```{r, eval = FALSE}
```

###4. What proportion of these have p-values <0.05 ?
```{r, eval = FALSE}
library(qvalue)
library(genefilter)
sex = sampleInfo$group
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = which(qvals<0.1 & !geneAnnotation$CHR%in%c("chrX","chrY"))
month = factor( format(sampleInfo$date,"%m"))
pval = rowttests(geneExpression[index,], month)$p.value
mean(pval<0.05)
```

The proportion of genes that are different are 0.8717949.

###5. The above result shows that the great majority of the autosomal genes show differences due to processing data. This provides further evidence that confounding is resulting in false positives. So we are going to try to model the month effect to better estimate the sex effect. We are going to use a linear model: Which of the following creates the appropriate design matrix? 
### A)X=model.matrix(∼sex+ethnicity)
### B)X=cbind(sex,as.numeric(month))
### C) It can’t be done with one line.
### D)X=model.matrix(∼sex+month)

D, X = model.matrix(~sex+month)

###6. How many of the q-values for the group comparison are now <0.1?
```{r, eval = FALSE}
sex = sampleInfo$group
month = factor( format(sampleInfo$date,"%m"))
X = model.matrix(~sex+month)
i = 234
y = geneExpression[i,]
fit = lm(y~X)
summary(fit)$coef

pvals <- t( sapply(1:nrow(geneExpression),function(j){
  y <- geneExpression[j,]
  fit <- lm(y~X-1)
  summary(fit)$coef[2,4] 
} ) )
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

17 of the q values are now less than 0.1.

###7. With this new list, what proportion of these are chrX and chrY?
```{r, eval = FALSE}
library(qvalue)
library(genefilter)
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

0.04763407 are chrX and chrY.

###8. How many on Y or X?
```{r, eval = FALSE}
```

###9. Now from the linear model above, extract the p-values related to the coefficient representing the October versus June differences using the same linear model. How many of the q-values for the month comparison are now <0.1?
```{r, eval = FALSE}
X = model.matrix(~sex+month)
pvals <- t( sapply(1:nrow(geneExpression),function(j){
  y <- geneExpression[j,]
  fit <- lm(y~X-1)
  summary(fit)$coef[3,4] # third coef, for month
} ) )
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

3170 of the q values are now less than 0.1.

##p448

###1. 
The answer is B. 

###2. 
The answer is C.

###3. 
```{r}
library(Biobase)
library(GSE5859Subset)
data(GSE5859Subset)

y = geneExpression - rowMeans(geneExpression)

library(RColorBrewer)
library(rafalib)
mypar(1,2)
n=ncol(y)
cors=cor(y)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
o <- order(sampleInfo$date)
yo<-y[,o]
cors=cor(yo)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))

library(rafalib)
sex = sampleInfo$group
mypar(1,2)
cors = cor(y)
image(cors)
o = order(sampleInfo$date)
image(cors[o,o])
```

###4.

```{r}
s <- svd(y)
What <- s$v[,1:2]
rownames(What)<-colnames(y)
round(What,2)
fit = s$u[,1:2]%*% (s$d[1:2]*t(What))
var(as.vector(fit))/var(as.vector(y))
```

###5.

```{r}
month=format(sampleInfo$date, "%m")
pcs = svd(y)$v[,1:2]
o = order(sampleInfo$date)
cols = as.numeric(month)[o]
mypar(2,1)
for(i in 1:2){
  plot(pcs[o,i],col=cols,xaxt="n",xlab="")
  label = gsub("2005-","",sampleInfo$date[o])
  axis(1,1:ncol(y),label,las=2)}
```

The answer is A.

###6.
```{r}
s = svd(y)
varexplained = s$d^2/ sum(s$d^2)
plot(varexplained)
sum(varexplained>0.10)
```

###7 + 8.
```{r}
s = svd(y)
month = factor( format(sampleInfo$date,"%m"))
cors = cor( as.numeric(month),s$v)
plot(t(cors))
which.max(abs(cors))
max(abs(cors))
```

###9 + 10.
```{r}
s = svd(y)
sex = sampleInfo$group
cors = cor( as.numeric(sex),s$v)
plot(t(cors))
which.max(abs(cors))
max(abs(cors))
```

###11.
```{r}
X <- model.matrix(~sex+s$v[,1:2])

library(limma)
X <- model.matrix(~sex+s$v[,1:2])
fit <- lmFit(y,X)
k <- 2 ##second coef, for month
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

###12.
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

###Exercises on page 459

###1.
```{r}
library("sva")
s <- svd(geneExpression-rowMeans(geneExpression))
cor(sampleInfo$group,s$v[,1])

sex = sampleInfo$group
mod = model.matrix(~sex)
svafit = sva(geneExpression,mod)
head(svafit$sv)

for(i in 1:ncol(svafit$sv)){
  print( cor(s$v[,i],svafit$sv[,i]) )
}

svaX<-model.matrix(~sex+svafit$sv)
library(limma)
fit <- lmFit(geneExpression,svaX)
k <- 2 #second coef, for sex
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

###2.
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)

res = sapply(1:nrow(geneExpression),function(i){
	y = geneExpression[i,]
	fit = lm(y~svaX-1)
	summary(fit)$coef[2,c(1,4)]
})
qvals = qvalue(res[2,])$qvalue
pcutoff = max( res[2,qvals < .1] )
library(rafalib)
mypar(1,1)
plot(res[1,],-log10(res[2,]),xlab="M",ylab="log10 p-value")
ind = which(geneAnnotation$CHR=="chrY")
points(res[1,ind],-log10(res[2,ind]),col=1,pch=16)
ind = which(geneAnnotation$CHR=="chrX")
points(res[1,ind],-log10(res[2,ind]),col=2,pch=16)
abline(h=-log10(pcutoff))
legend("bottomleft",c("chrX","chrY"),col=c(2,1),pch=16)
```

